backend/src/main/java/com/javahelp/backend/data/ISurveyResponseStore.java

```diff
@@ -0,0 +1,67 @@
+package com.javahelp.backend.data;
+
+import com.amazonaws.regions.Regions;
+import com.javahelp.model.survey.SurveyResponse;
+import com.javahelp.model.user.User;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public interface ISurveyResponseStore {
+    /**
+     * @return Default implementation of {@link ISurveyResponseStore}
+     */
+    static ISurveyResponseStore getDefaultImplementation() {
+        return new DynamoDBSurveyResponseStore("javahelpBackendFilledSurveys", Regions.US_EAST_1);
+    }
+
+    /**
+     * Creates a {@link SurveyResponse}
+     *
+     * @param userID {@link String} id of author (user) of this surveyResponse
+     * @param surveyResponse {@link SurveyResponse} to create
+     * @return {@link SurveyResponse} that was created
+     */
+    SurveyResponse create(String userID, SurveyResponse surveyResponse);
+
+    /**
+     * @param id {@link String} id of the {@link SurveyResponse}
+     * @return {@link SurveyResponse} with the specified id
+     */
+    SurveyResponse read(String id);
+
+    /**
+     * Deletes the specified {@link SurveyResponse}
+     *
+     * @param id {@link String} id of the {@link SurveyResponse} to delete
+     */
+    void delete(String id);
+
+    /**
+     * Removes all {@link SurveyResponse}s in database.
+     * ONLY use during preliminary testing!
+     */
+    void cleanTable();
+
+    /**
+     *
+     * @param userID {@link String} id of the user to be queried.
+     * @return {@link List <SurveyResponse>} of the specified user.
+     */
+    List<SurveyResponse> readByUser(String userID);
+
+    /**
+     *
+     * @param constraint {@link Map <>} which specifies the required attributes
+     *                   from the User.
+     * @return {@link Set <User>} with the specified constraints.
+     */
+    Set<String> readByConstraint(Map<String, Set<String>> constraint);
```

Comment:
I agree, but I also think the current specified type for the constraint is quite universal as it simply specifies the set of possible solutions to a given query constraint. How that query constraint is satisfied is not specified in the interface. Do correct me if I'm wrong

Reply:
This is true; however the way the input is currently setup, arbitrary strings will have to be parsed into corresponding field/column names depending on the underlying implementation. Additionally, the constraints will need to be parsed from strings into queries for the database. This essentially amounts to specifying our own query language, and then implementing that query language over some underlying database. This adds a lot of complexity, and since the method takes in strings, makes future refactors much more difficult. While this is definitely a functioning way to do this, it could lead to future issues. Either way, although there are better ways to do it, I think it's good enough for our purposes. If we really want to we can mark it as an area for future improvement in our submission,